src/components/RegistrationForm/RegistrationForm.tsx

```diff
@@ -0,0 +1,198 @@
+import { useState, FormEvent, ChangeEvent, useMemo } from 'react';
+import { Link } from 'react-router-dom';
+
+import { FormInput } from '../FormInput/FormInput';
+import style from './RegistrationForm.module.scss';
+
+type FormData = {
+  firstName: string;
+  lastName: string;
+  email: string;
+  password: string;
+  confirmPassword: string;
+};
+
+type TouchedFields = {
+  firstName: boolean;
+  lastName: boolean;
+  email: boolean;
+  password: boolean;
+  confirmPassword: boolean;
+};
+
+const INITIAL_FORM_DATA: FormData = {
+  firstName: '',
+  lastName: '',
+  email: '',
+  password: '',
+  confirmPassword: '',
+};
+
+const INITIAL_TOUCHED: TouchedFields = {
+  firstName: false,
+  lastName: false,
+  email: false,
+  password: false,
+  confirmPassword: false,
+};
+
+const ERROR_MESSAGES = {
+  firstName: 'First name cannot be empty or contain numbers',
+  lastName: 'Last name cannot be empty or contain numbers',
+  email: 'Please enter a valid email',
+  password:
+    'Password must have at least 8 characters, 1 uppercase letter and 1 number',
+  passwordMatch: "Passwords don't match",
+};
+
+const validators = {
+  name: (value: string): boolean => /^[^0-9]+$/.test(value) && value.length > 0,
+  email: (value: string): boolean => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value),
+  password: (value: string): boolean =>
+    /^(?=.*[A-Z])(?=.*\d).{8,}$/.test(value),
+};
+
+export const RegistrationForm = () => {
+  const [formData, setFormData] = useState<FormData>(INITIAL_FORM_DATA);
+  const [touched, setTouched] = useState<TouchedFields>(INITIAL_TOUCHED);
+  const [shakeAnimation, setShakeAnimation] = useState(false);
+  const [isSubmitted, setIsSubmitted] = useState(false);
+
+  const formErrors = useMemo(
+    () => ({
+      firstName: touched.firstName && !validators.name(formData.firstName),
+      lastName: touched.lastName && !validators.name(formData.lastName),
+      email: touched.email && !validators.email(formData.email),
+      password: touched.password && !validators.password(formData.password),
+      confirmPassword:
+        touched.confirmPassword &&
+        formData.password !== formData.confirmPassword,
+    }),
+    [formData, touched],
+  );
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    setFormData((prev) => ({ ...prev, [name]: value }));
+  };
+
+  const handleBlur = (fieldName: keyof TouchedFields) => {
+    setTouched((prev) => ({ ...prev, [fieldName]: true }));
+  };
```

Comment:
If the validations is in handleChange, the whole touched system is not needed. onBlurs won't be needed as well.

KISS - KEEP IT SIMPLE, STUPID